nomos-da/network/core/src/protocols/replication/behaviour_v2.rs

```diff
@@ -0,0 +1,677 @@
+    outbound_tasks: FuturesUnordered<OutboundTask>,
+    /// Seen messages cache holds a record of seen messages
+    seen_message_cache: IndexSet<(Vec<u8>, SubnetworkId)>,
+    /// Waker that handles polling
+    waker: Option<Waker>,
+}
+
+impl<Membership> ReplicationBehaviour<Membership> {
+    pub fn new(peer_id: PeerId, membership: Membership) -> Self {
+        let stream_behaviour = libp2p_stream::Behaviour::new();
+        let mut control = stream_behaviour.new_control();
+        let incoming_streams = control
+            .accept(REPLICATION_PROTOCOL)
+            .expect("A unique protocol can be accepted only once");
+        let incoming_tasks = FuturesUnordered::new();
+        let outbound_tasks = FuturesUnordered::new();
+
+        Self {
+            local_peer_id: peer_id,
+            stream_behaviour,
+            control,
+            incoming_streams,
+            incoming_tasks,
+            membership,
+            connected: HashSet::default(),
+            pending_outbound: PendingOutbound::default(),
+            outbound_streams: HashMap::default(),
+            outbound_tasks,
+            seen_message_cache: IndexSet::default(),
+            waker: None,
+        }
+    }
+
+    pub fn update_membership(&mut self, membership: Membership) {
+        self.membership = membership;
+    }
+}
+
+impl<M> ReplicationBehaviour<M>
+where
+    M: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId>,
+{
+    /// Check if some peer membership lies in at least a single subnetwork that
+    /// the local peer is a member too.
+    fn is_neighbour(&self, peer_id: &PeerId) -> bool {
+        self.membership
+            .membership(&self.local_peer_id)
+            .intersection(&self.membership.membership(peer_id))
+            .count()
+            > 0
+    }
+
+    fn no_loopback_member_peers_of(&self, subnetwork: SubnetworkId) -> HashSet<PeerId> {
+        let mut peers = self.membership.members_of(&subnetwork);
+        // no loopback
+        peers.remove(&self.local_peer_id);
+        peers
+    }
+
+    fn replicate_message(&mut self, message: &DaMessage) {
+        let message_id = (message.blob.blob_id.to_vec(), message.subnetwork_id);
+        if self.seen_message_cache.contains(&message_id) {
+            return;
+        }
+        self.seen_message_cache.insert(message_id);
+        self.send_message(message);
+    }
+
+    pub fn send_message(&mut self, message: &DaMessage) {
+        // Push a message in the queue for every single peer connected that is a member
+        // of the selected subnetwork_id
+        let peers = self.no_loopback_member_peers_of(message.subnetwork_id);
+
+        self.connected
+            .iter()
+            .filter(|peer_id| peers.contains(peer_id))
+            .for_each(|peer_id| {
+                self.pending_outbound
+                    .enqueue_message(*peer_id, message.clone());
+            });
+
+        self.try_wake();
```

Comment:
Since `send_message` and `replicate_messge` are being called in poll only, we could just pass the `cx` and wake by ref instead of storing the waker. Storing waker in such network behaviours is a good solution when we want to trigger polling when receiving a new swarm event or something is needed to be reevaluated on new connection.